experimental/dds/tree2/src/feature-libraries/editable-tree-2/editableTreeTypes.ts

```diff
@@ -0,0 +1,493 @@
+ * See {@link StructTyped} for the schema aware extensions to this.
+ *
+ * @alpha
+ */
+export interface Struct extends UntypedTree {
+	/**
+	 * {@link LocalNodeKey} that identifies this node.
+	 */
+	readonly localNodeKey?: LocalNodeKey;
+}
+
+/**
+ * Leaf holding a value.
+ *
+ * @remarks
+ * Leaves are immutable and have no children.
+ * Leaf unboxes its content, so in schema aware APIs which do unboxing, the Leaf itself will be skipped over and its value will be returned directly.
+ * @alpha
+ */
+export interface Leaf<TSchema extends LeafSchema> extends UntypedTree {
+	/**
+	 * Value stored on this node.
+	 */
+	readonly value: SchemaAware.InternalTypes.TypedValue<TSchema["leafValue"]>;
+}
+
+/**
+ * A node that behaves like struct, providing properties to access its fields.
+ *
+ * @alpha
+ */
+export type StructTyped<TSchema extends StructSchema> = Struct &
+	StructFields<TSchema["structFieldsObject"]>;
+
+/**
+ * Properties to access a struct nodes fields. See {@link StructTyped}.
+ *
+ * @privateRemarks
+ * TODO: support custom field keys
+ *
+ * @alpha
+ */
+export type StructFields<TFields extends RestrictiveReadonlyRecord<string, FieldSchema>> =
+	// Getters
+	{
+		readonly [key in keyof TFields]: UnboxField<TFields[key]>;
+	} & {
+		readonly // boxed fields (TODO: maybe remove these when same as non-boxed version?)
+		[key in keyof TFields as `boxed${Capitalize<key & string>}`]: TypedField<TFields[key]>;
+	};
+// TODO: Add `set` method when FieldKind provides a setter (and derive the type from it).
+// set(key: FieldKey, content: FlexibleFieldContent<TSchema["mapFields"]>): void;
+// {
+// 	readonly [key in keyof TFields as `set${Capitalize<key & string>}`]: (
+// 		content: FlexibleFieldContent<TFields[key]>,
+// 	) => void;
+// };
+// This could be enabled to allow assignment via `=` in some cases.
+// & {
+// 	// Setter properties (when the type system permits)
+// 	[key in keyof TFields]: UnwrappedField<TFields[key]> & StructSetContent<TFields[key]>;
+// }
+
+// #endregion
+
+// #region Field Kinds
+
+/**
+ * Strongly typed tree literals for inserting as the content of a field.
+ * @alpha
+ */
+export type FlexibleFieldContent<TSchema extends FieldSchema> = SchemaAware.TypedField<
+	TSchema,
+	SchemaAware.ApiMode.Flexible
+>;
+
+/**
+ * Strongly typed tree literals for inserting as a node.
+ * @alpha
+ */
+export type FlexibleNodeContent<TTypes extends AllowedTypes> = SchemaAware.AllowedTypesToTypedTrees<
+	SchemaAware.ApiMode.Flexible,
+	TTypes
+>;
+
+/**
+ * Field this stores a sequence of children.
+ *
+ * @remarks
+ * Allows for concurrent editing based on index, adjusting the locations of indexes as needed so they apply to the same logical place in the sequence when rebased and merged.
+ * @alpha
+ */
+export interface Sequence<TTypes extends AllowedTypes> extends UntypedField {
+	/**
+	 * Gets a node of this field by its index with unboxing.
+	 * Note that a node must exist at the given index.
```

Comment:
Do we have a contract for what will happen if this invariant is violated?

Reply:
No. We do not yet have a policy how to handle API invariants as far as I know, but there is work planned to get a decision on it.

Generally if you use the reference forest implementation (configured when setting up the factory) you tend to get asserts for a lot of invalid use, but its not really robust, and may or may not help this case.